File: frontend/src/types/ai.ts
import type { ReportPreset } from "./reports";

export interface DataRange {
  from: string;
  to: string;
}

export interface ForecastDay {
  date: string;
  predicted_quantity: number;
  low: number;
  high: number;
  actual_quantity: number | null;
  is_sufficient_history: boolean;
}

export interface DemandForecast {
  route_id: number | null;
  milk_type_id: number | null;
  horizon_days: number;
  date_from: string;
  date_to: string;
  method: string;
  is_sufficient_history: boolean;
  message: string | null;
  total_expected: number | null;
  low_range: number | null;
  high_range: number | null;
  items: ForecastDay[];
}

export type AnomalySeverity = "HIGH" | "MEDIUM" | "LOW";

export interface AnomalyItem {
  type: string;
  severity: AnomalySeverity;
  title: string;
  description: string;
  entity_type: "session" | "route" | "customer" | "payment";
  entity_id: number;
  entity_name: string;
  metric: string;
  expected: number;
  actual: number;
  deviation: number;
  occurred_on: string;
  suggested_action: string;
}

export interface AnomalyReport {
  generated_at: string;
  count: number;
  items: AnomalyItem[];
}

export type RiskLevel = "LOW" | "MEDIUM" | "HIGH";

export interface ChurnFactor {
  factor: string;
  weight: number;
  contribution: number;
}

export interface ChurnRiskItem {
  customer_id: number;
  customer_code: string;
  customer_name: string;
  route_name: string;
  risk_score: number;
  risk_level: RiskLevel;
  factors: ChurnFactor[];
  suggested_action: string;
}

export interface ChurnRiskReport {
  generated_at: string;
  count: number;
  items: ChurnRiskItem[];
}

export interface AIInsightsResponse {
  generated_at: string;
  stats_only: boolean;
  data_range: DataRange;
  narrative: string | null;
  operational: Record<string, unknown>;
  forecast: DemandForecast;
  anomalies: AnomalyReport;
  churn_risk: ChurnRiskReport;
}

export interface ChatMessage {
  role: "user" | "assistant";
  content: string;
}

export interface ChatRequest {
  message: string;
  history: ChatMessage[];
}

export interface ChatResponse {
  reply: string;
  data_range: DataRange;
  sources: string[];
  stats_only: boolean;
}

export interface ForecastParams {
  route_id?: number;
  milk_type_id?: number;
  horizon_days?: number;
  refresh?: boolean;
}

export interface AnomalyParams {
  route_id?: number;
  days_back?: number;
  refresh?: boolean;
}

export interface ChurnRiskParams {
  route_id?: number;
  limit?: number;
  refresh?: boolean;
}

export interface InsightsParams {
  preset?: ReportPreset;
  from_date?: string;
  to_date?: string;
  refresh?: boolean;
}

File: app/schemas/ai.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field, ConfigDict


class DataRange(BaseModel):
    from_: date = Field(alias="from")
    to: date

    model_config = ConfigDict(populate_by_name=True)


class ForecastDay(BaseModel):
    date: date
    predicted_quantity: float
    low: float
    high: float
    actual_quantity: float | None = None
    is_sufficient_history: bool


class DemandForecast(BaseModel):
    route_id: int | None = None
    milk_type_id: int | None = None
    horizon_days: int
    date_from: date
    date_to: date
    method: str = "weekday_seasonal_moving_average"
    is_sufficient_history: bool
    message: str | None = None
    total_expected: float | None = None
    low_range: float | None = None
    high_range: float | None = None
    items: list[ForecastDay]


class AnomalyItem(BaseModel):
    type: str
    severity: Literal["HIGH", "MEDIUM", "LOW"]
    title: str
    description: str
    entity_type: Literal["session", "route", "customer", "payment"]
    entity_id: int
    entity_name: str
    metric: str
    expected: float
    actual: float
    deviation: float
    occurred_on: date
    suggested_action: str


class AnomalyReport(BaseModel):
    generated_at: datetime
    count: int
    items: list[AnomalyItem]


class ChurnFactor(BaseModel):
    factor: str
    weight: int
    contribution: int


class ChurnRiskItem(BaseModel):
    customer_id: int
    customer_code: str
    customer_name: str
    route_name: str
    risk_score: int
    risk_level: Literal["LOW", "MEDIUM", "HIGH"]
    factors: list[ChurnFactor]
    suggested_action: str


class ChurnRiskReport(BaseModel):
    generated_at: datetime
    count: int
    items: list[ChurnRiskItem]


class AIInsightsResponse(BaseModel):
    generated_at: datetime
    stats_only: bool
    data_range: DataRange
    narrative: str | None = None
    operational: dict
    forecast: DemandForecast
    anomalies: AnomalyReport
    churn_risk: ChurnRiskReport


class ChatMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str


class ChatRequest(BaseModel):
    message: str = Field(min_length=1, max_length=2000)
    history: list[ChatMessage] = Field(default_factory=list, max_length=8)


class ChatResponse(BaseModel):
    reply: str
    data_range: DataRange
    sources: list[str]
    stats_only: bool = False
